packages/block-library/src/table/state.js

```diff
@@ -26,105 +28,186 @@ export function createTable( {
+ * @param {Object}   selection  The selection of cells to update.
+ * @param {Function} updateCell A function to update the selected cell attributes.
+ *
+ * @return {Object} New table state including the updated cells.
+ */
+export function updateSelectedCell( state, selection, updateCell ) {
+	if ( ! selection ) {
+		return state;
+	}
+
+	const tableSections = pick( state, [ 'head', 'body', 'foot' ] );
+	const {
+		sectionName: selectionSectionName,
+		rowIndex: selectionRowIndex,
+	} = selection;
+
+	return mapValues( tableSections, ( section, sectionName ) => {
+		if ( selectionSectionName && selectionSectionName !== sectionName ) {
+			return section;
+		}
+
+		return section.map( ( row, rowIndex ) => {
+			if ( selectionRowIndex && selectionRowIndex !== rowIndex ) {
 				return row;
 			}
 
 			return {
-				cells: row.cells.map( ( cell, currentColumnIndex ) => {
-					if ( currentColumnIndex !== columnIndex ) {
-						return cell;
+				cells: row.cells.map( ( cellAttributes, columnIndex ) => {
+					const cellLocation = {
+						sectionName,
+						columnIndex,
+						rowIndex,
+					};
+
+					if ( ! isCellSelected( cellLocation, selection ) ) {
+						return cellAttributes;
 					}
 
-					return {
-						...cell,
-						content,
-					};
+					return updateCell( cellAttributes );
 				} ),
 			};
-		} ),
-	};
+		} );
+	} );
+}
+
+/**
+ * Returns whether the cell at `cellLocation` is included in the selection `selection`.
+ *
+ * @param {Object} cellLocation An object containing cell location properties.
+ * @param {Object} selection    An object containing selection properties.
+ *
+ * @return {boolean} True if the cell is selected, false otherwise.
+ */
+export function isCellSelected( cellLocation, selection ) {
+	if ( ! cellLocation || ! selection ) {
+		return false;
+	}
+
+	switch ( selection.type ) {
+		case 'column':
+			return selection.type === 'column' && cellLocation.columnIndex === selection.columnIndex;
+		case 'cell':
+			return selection.type === 'cell' &&
+				cellLocation.sectionName === selection.sectionName &&
+				cellLocation.columnIndex === selection.columnIndex &&
+				cellLocation.rowIndex === selection.rowIndex;
+	}
 }
 
 /**
  * Inserts a row in the table state.
  *
- * @param {Object} state            Current table state.
- * @param {string} options.section  Section in which to insert the row.
- * @param {number} options.rowIndex Row index at which to insert the row.
+ * @param {Object} state                Current table state.
+ * @param {string} options.sectionName  Section in which to insert the row.
+ * @param {number} options.rowIndex     Row index at which to insert the row.
  *
  * @return {Object} New table state.
  */
 export function insertRow( state, {
-	section,
+	sectionName,
 	rowIndex,
 	columnCount,
 } ) {
-	const cellCount = columnCount || state[ section ][ 0 ].cells.length;
+	const firstRow = getFirstRow( state );
+	const cellCount = columnCount || get( firstRow, [ 'cells', 'length' ], 2 );
```

Comment:
Why does it default to 2?

Reply:
That's a fair comment. I thought `2` because that's what the table block defaults to when created.

Maybe this should instead be considered unexpected and the `insertRow` function should return early with no effect if `cellCount` is undefined.

Reply:
I changed it to return early with unchanged state if `cellCount` is unknown.

Reply:
I think so too. If there's no amount of rows, there's nothing to be done? If you'd like to keep a default, I feel like it shouldn't be hardcoded here, but set by the component instead (which knows the default row count).